bigtop-packages/src/charm/hadoop/layer-hadoop-namenode/reactive/namenode.py

```diff
@@ -0,0 +1,206 @@
+        props['dfs.namenode.http-bind-host'] = '0.0.0.0'
+        props['dfs.namenode.https-bind-host'] = '0.0.0.0'
+
+    # We need to create the 'mapred' user/group since we are not installing
+    # hadoop-mapreduce. This is needed so the namenode can access yarn
+    # job history files in hdfs. Also add our ubuntu user to the hadoop
+    # and mapred groups.
+    get_layer_opts().add_users()
+
+    set_state('apache-bigtop-namenode.installed')
+    hookenv.status_set('maintenance', 'namenode installed')
+
+
+@when('apache-bigtop-namenode.installed')
+@when_not('apache-bigtop-namenode.started')
+def start_namenode():
+    hookenv.status_set('maintenance', 'starting namenode')
+    # NB: service should be started by install, but this may be handy in case
+    # we have something that removes the .started state in the future. Also
+    # note we restart here in case we modify conf between install and now.
+    host.service_restart('hadoop-hdfs-namenode')
+    for port in get_layer_opts().exposed_ports('namenode'):
+        hookenv.open_port(port)
+    set_state('apache-bigtop-namenode.started')
+    hookenv.status_set('maintenance', 'namenode started')
+
+
+###############################################################################
+# Slave methods
+###############################################################################
+@when('datanode.joined')
+@when_not('apache-bigtop-namenode.installed')
+def send_dn_install_info(datanode):
+    """Send datanodes enough relation data to start their install."""
+    send_early_install_info(datanode)
+
+
+@when('apache-bigtop-namenode.started', 'datanode.joined')
+def send_dn_all_info(datanode):
+    """Send datanodes all dfs-slave relation data.
+
+    At this point, the namenode is ready to serve datanodes. Send all
+    dfs-slave relation data so that our 'namenode.ready' state becomes set.
+    """
+    bigtop = Bigtop()
+    fqdn = get_fqdn()
+    hdfs_port = get_layer_opts().port('namenode')
+    webhdfs_port = get_layer_opts().port('nn_webapp_http')
+
+    datanode.send_spec(bigtop.spec())
+    datanode.send_namenodes([fqdn])
+    datanode.send_ports(hdfs_port, webhdfs_port)
+
+    # hosts_map, ssh_key, and clustername are required by the dfs-slave
+    # interface to signify NN's readiness. Send them, even though they are not
+    # utilized by bigtop.
+    # NB: update KV hosts with all datanodes prior to sending the hosts_map
+    # because dfs-slave gates readiness on a DN's presence in the hosts_map.
+    utils.update_kv_hosts(datanode.hosts_map())
+    datanode.send_hosts_map(utils.get_kv_hosts())
+    datanode.send_ssh_key('invalid')
+    datanode.send_clustername(hookenv.service_name())
+
+    # update status with slave count and report ready for hdfs
+    num_slaves = len(datanode.nodes())
+    hookenv.status_set('active', 'ready ({count} datanode{s})'.format(
+        count=num_slaves,
+        s='s' if num_slaves > 1 else '',
```

Comment:
Nice use of a ternary operator. I always am tempted to sneak them in, and they're usually not warranted. But this makes the code concise, and still feels basically Pythonic.